### PR TITLE
modernize `RecoVertex/BeamSpotProducer`

### LINKED ISSUE #24: Deprecated EDAnalyzer in BeamSpotProducer
A recent PR touched on the BeamSpotProducer and so the Static Analyzer brought up a warning that the BeamSpotProducer is using a Deprecated EDAnalyzer
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23a921/18400/llvm-analysis/
More info about the legacy EDAnalyzer can be found in [*].

[*] https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedAnalysisEDAnalyzer

### PR DESCRIPTION
**Pull request by mmusich**
fixes https://github.com/cms-AlCaDB/AlCaTools/issues/24

#### PR description:

Modernize `RecoVertex/BeamSpotProducer`:
   * merge headers with plugin implementation
   * transformed legacy `EDAnalyzer`s into `edm::one::EDAnalyzer`s
   * move `cout` to thread-safe `MessageLogger` instructions
   * remove commented out code
   * add `fillDescriptions` methods  

#### PR validation:

It compiles, also tested with unit tests.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

N/A

### CODE REVIEW

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc [hunk 1]
```diff
@@ -82,12 +111,10 @@ void BeamSpotAnalyzer::beginLuminosityBlock(const edm::LuminosityBlock& lumiSeg,
...
-  //LogDebug("BeamSpotAnalyzer") <<
-  std::cout << "for lumis " << beginLumiOfBSFit_ << " - " << endLumiOfBSFit_ << std::endl
-            << "number of selected tracks = " << theBeamFitter->getNTracks() << std::endl;
-  std::cout << "number of selected PVs = " << theBeamFitter->getNPVs() << std::endl;
-  //std::cout << "number of selected PVs per bx: " << std::endl;
-  //theBeamFitter->getNPVsperBX();
+  edm::LogPrint("BeamSpotAnalyzer") << "for lumis " << beginLumiOfBSFit_ << " - " << endLumiOfBSFit_ << std::endl
+                                    << "number of selected tracks = " << theBeamFitter->getNTracks() << std::endl;
+  edm::LogPrint("BeamSpotAnalyzer") << "number of selected PVs = " << theBeamFitter->getNPVs() << std::endl;
+  //edm::LogPrint("BeamSpotAnalyzer") << "number of selected PVs per bx: " << theBeamFitter->getNPVsperBX() std::endl;
```
Comment (tvami) [inline RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc:117]: Is the `std::endl` needed when `edm::LogPrint` is used? I though the LogPrint takes care of the new line.
Also should the commented out line be removed?

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc [hunk 2]
```diff
@@ -110,61 +137,62 @@ void BeamSpotAnalyzer::endLuminosityBlock(const edm::LuminosityBlock& lumiSeg, c
...
+    edm::LogPrint("BeamSpotAnalyzer") << " for runs: " << ftmprun0 << " - " << ftmprun << std::endl;
+    edm::LogPrint("BeamSpotAnalyzer") << " for lumi blocks : " << LSRange.first << " - " << LSRange.second << std::endl;
+    edm::LogPrint("BeamSpotAnalyzer") << " lumi counter # " << countLumi_ << std::endl;
+    edm::LogPrint("BeamSpotAnalyzer") << bs << std::endl;
+    edm::LogPrint("BeamSpotAnalyzer") << "[BeamFitter] fit failed \n" << std::endl;
+    // accumulate more events
+    // disable this for the moment
     //resetFitNLumi_ += 1;
-    //std::cout << "reset fitNLumi " << resetFitNLumi_ << std::endl;
+    //edm::LogPrint("BeamSpotAnalyzer") << "reset fitNLumi " << resetFitNLumi_ << std::endl;
```
Comment (tvami) [inline RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc:158]: Is the std::endl needed when edm::LogPrint is used? I though the LogPrint takes care of the new line.
Also should the commented out lines be removed?

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc [hunk 3]
```diff
@@ -110,61 +137,62 @@ void BeamSpotAnalyzer::endLuminosityBlock(const edm::LuminosityBlock& lumiSeg, c
...
   }
 
   if (resetFitNLumi_ > 0 && countLumi_ % resetFitNLumi_ == 0) {
     std::vector<BSTrkParameters> theBSvector = theBeamFitter->getBSvector();
-    std::cout << "Total number of tracks accumulated = " << theBSvector.size() << std::endl;
-    std::cout << "Reset track collection for beam fit" << std::endl;
+    edm::LogPrint("BeamSpotAnalyzer") << "Total number of tracks accumulated = " << theBSvector.size() << std::endl;
+    edm::LogPrint("BeamSpotAnalyzer") << "Reset track collection for beam fit" << std::endl;
```
Comment (tvami) [inline RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc:164]: Is the std::endl needed when edm::LogPrint is used? I though the LogPrint takes care of the new line.

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc [hunk 4]
```diff
@@ -110,61 +137,62 @@ void BeamSpotAnalyzer::endLuminosityBlock(const edm::LuminosityBlock& lumiSeg, c
...
   }
 }
 
 void BeamSpotAnalyzer::endJob() {
-  std::cout << "\n-------------------------------------\n" << std::endl;
-  std::cout << "\n Total number of events processed: " << ftotalevents << std::endl;
-  std::cout << "\n-------------------------------------\n\n" << std::endl;
+  edm::LogPrint("BeamSpotAnalyzer") << "\n-------------------------------------\n" << std::endl;
+  edm::LogPrint("BeamSpotAnalyzer") << "\n Total number of events processed: " << ftotalevents << std::endl;
+  edm::LogPrint("BeamSpotAnalyzer") << "\n-------------------------------------\n\n" << std::endl;
```
Comment (tvami) [inline RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc:179]: Is the std::endl needed when edm::LogPrint is used? I thought the LogPrint takes care of the new line.

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc [hunk 5]
```diff
@@ -110,61 +137,62 @@ void BeamSpotAnalyzer::endLuminosityBlock(const edm::LuminosityBlock& lumiSeg, c
...
+      edm::LogPrint("BeamSpotAnalyzer") << " for lumi blocks : " << LSRange.first << " - " << LSRange.second
+                                        << std::endl;
+      edm::LogPrint("BeamSpotAnalyzer") << " lumi counter # " << countLumi_ << std::endl;
+      edm::LogPrint("BeamSpotAnalyzer") << beam_default << std::endl;
 
       if (write2DB_) {
-        std::cout << "\n-------------------------------------\n\n" << std::endl;
-        std::cout << " write results to DB..." << std::endl;
+        edm::LogPrint("BeamSpotAnalyzer") << "\n-------------------------------------\n\n" << std::endl;
+        edm::LogPrint("BeamSpotAnalyzer") << " write results to DB..." << std::endl;
```
Comment (tvami) [inline RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc:195]: Same comment for the `std::endl`

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc [hunk 6]
```diff
@@ -175,14 +203,28 @@ void BeamSpotAnalyzer::endJob() {
     if ((runbeamwidthfit_)) {
       theBeamFitter->runBeamWidthFitter();
       reco::BeamSpot beam_width = theBeamFitter->getBeamWidth();
-      std::cout << beam_width << std::endl;
+      edm::LogPrint("BeamSpotAnalyzer") << beam_width << std::endl;
+    } else {
+      edm::LogPrint("BeamSpotAnalyzer") << "[BeamSpotAnalyzer] beamfit fails !!!" << std::endl;
```
Comment (tvami) [inline RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc:208]: Same comment for the `std::endl`

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotFromDB.cc [hunk 7]
```diff
@@ -1,47 +1,49 @@
...
-
-  std::cout << " for runs: " << iEvent.id().run() << " - " << iEvent.id().run() << std::endl;
-  //std::cout << iEvent.getRun().beginTime().value() << std::endl;
-  //std::cout << iEvent.time().value() << std::endl;
-  std::cout << *mybeamspot << std::endl;
+  const BeamSpotObjects* mybeamspot = &iSetup.getData(m_beamToken);
+  edm::LogPrint("BeamSpotFromDB") << " for runs: " << iEvent.id().run() << " - " << iEvent.id().run() << std::endl;
+  //edm::LogPrint("BeamSpotFromDB") << iEvent.getRun().beginTime().value() << std::endl;
+  //edm::LogPrint("BeamSpotFromDB") << iEvent.time().value() << std::endl;
+  edm::LogPrint("BeamSpotFromDB") << *mybeamspot << std::endl;
```
Comment (tvami) [inline RecoVertex/BeamSpotProducer/plugins/BeamSpotFromDB.cc:45]: Same comment for the `std::endl` + commented out code

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotWrite2DB.cc [hunk 8]
```diff
@@ -1,53 +1,75 @@
...
+  edm::LogInfo("BeamSpotWrite2DB")
+      << " Content of the file is expected to have this format with the first column as a keyword:" << std::endl;
+  edm::LogInfo("BeamSpotWrite2DB") << " x\n y\n z\n sigmaZ\n dxdz\n dydz\n beamWidthX\n beamWidthY" << std::endl;
+  for (int i = 0; i < 7; i++) {
+    for (int j = 0; j < 7; j++) {
+      edm::LogInfo("BeamSpotWrite2DB") << " cov[" << i << "][" << j << "] cov[" << i << "][" << j << "] cov[" << i
+                                       << "][" << j << "] cov[" << i << "][" << j << "] cov[" << i << "][" << j
+                                       << "] cov[" << j << "][" << j << "] cov[" << i << "][" << j << "]" << std::endl;
+    }
+  }
```
Comment (tvami) [inline RecoVertex/BeamSpotProducer/plugins/BeamSpotWrite2DB.cc:72]: Same comment as above for the `std::endl`

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by slava77** [comment]
@cmsbuild please test

CMSSW_12_1_X_2021-09-11-1100 is out now (10-1100 was too old for this PR)

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by tvami** [comment]
+alca

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc [hunk 9]
```diff
@@ -82,12 +111,10 @@ void BeamSpotAnalyzer::beginLuminosityBlock(const edm::LuminosityBlock& lumiSeg,
...
-  //LogDebug("BeamSpotAnalyzer") <<
-  std::cout << "for lumis " << beginLumiOfBSFit_ << " - " << endLumiOfBSFit_ << std::endl
-            << "number of selected tracks = " << theBeamFitter->getNTracks() << std::endl;
-  std::cout << "number of selected PVs = " << theBeamFitter->getNPVs() << std::endl;
-  //std::cout << "number of selected PVs per bx: " << std::endl;
-  //theBeamFitter->getNPVsperBX();
+  edm::LogPrint("BeamSpotAnalyzer") << "for lumis " << beginLumiOfBSFit_ << " - " << endLumiOfBSFit_ << std::endl
+                                    << "number of selected tracks = " << theBeamFitter->getNTracks();
+  edm::LogPrint("BeamSpotAnalyzer") << "number of selected PVs = " << theBeamFitter->getNPVs();
+  //edm::LogPrint("BeamSpotAnalyzer") << "number of selected PVs per bx: " << theBeamFitter->getNPVsperBX() std::endl;
```
Comment (jpata) [inline RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc:117]:
```suggestion
```

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc [hunk 10]
```diff
@@ -110,61 +137,61 @@ void BeamSpotAnalyzer::endLuminosityBlock(const edm::LuminosityBlock& lumiSeg, c
...
-    // dissable this for the moment
+    edm::LogPrint("BeamSpotAnalyzer") << "\n Empty Beam spot fit";
+    edm::LogPrint("BeamSpotAnalyzer") << " for runs: " << ftmprun0 << " - " << ftmprun;
+    edm::LogPrint("BeamSpotAnalyzer") << " for lumi blocks : " << LSRange.first << " - " << LSRange.second;
+    edm::LogPrint("BeamSpotAnalyzer") << " lumi counter # " << countLumi_;
+    edm::LogPrint("BeamSpotAnalyzer") << bs;
+    edm::LogPrint("BeamSpotAnalyzer") << "[BeamFitter] fit failed \n";
+    // accumulate more events
+    // disable this for the moment
     //resetFitNLumi_ += 1;
```
Comment (jpata) [inline RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc:157]: should this be removed?

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc [hunk 11]
```diff
@@ -110,61 +137,61 @@ void BeamSpotAnalyzer::endLuminosityBlock(const edm::LuminosityBlock& lumiSeg, c
...
-    //std::cout << "reset fitNLumi " << resetFitNLumi_ << std::endl;
+    //edm::LogPrint("BeamSpotAnalyzer") << "reset fitNLumi " << resetFitNLumi_ ;
```
Comment (jpata) [inline RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc:158]:
```suggestion
```

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotFromDB.cc [hunk 12]
```diff
@@ -1,47 +1,49 @@
...
-  const BeamSpotObjects* mybeamspot = beamhandle.product();
-
-  std::cout << " for runs: " << iEvent.id().run() << " - " << iEvent.id().run() << std::endl;
-  //std::cout << iEvent.getRun().beginTime().value() << std::endl;
-  //std::cout << iEvent.time().value() << std::endl;
-  std::cout << *mybeamspot << std::endl;
+  const BeamSpotObjects* mybeamspot = &iSetup.getData(m_beamToken);
+  edm::LogPrint("BeamSpotFromDB") << " for runs: " << iEvent.id().run() << " - " << iEvent.id().run();
+  //edm::LogPrint("BeamSpotFromDB") << iEvent.getRun().beginTime().value();
+  //edm::LogPrint("BeamSpotFromDB") << iEvent.time().value();
```
Comment (jpata) [inline RecoVertex/BeamSpotProducer/plugins/BeamSpotFromDB.cc:44]:
```suggestion
```

**Review by jpata — COMMENTED** [review]
perhaps some more commented-out code can be removed here?

**Comment by mmusich** [comment]
@jpata I am not going to remove the commented lines as I am not the original author (nor the long term maintainer) of the code.

**Comment by jpata** [comment]
+reconstruction
- for 993a9a6
- no changes in reco, code changes are minimal

**Comment by perrotta** [comment]
+1

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc [hunk 13]
```diff
@@ -2,28 +2,60 @@
...
-
  author: Francisco Yumiceva, Fermilab (yumiceva@fnal.gov)
          Geng-Yuan Jeng, UC Riverside (Geng-Yuan.Jeng@cern.ch)
 
-
 ________________________________________________________________**/
 
 // C++ standard
 #include <string>
+#include <string>
```
[the remark this answers is not in the document]
Comment (mmusich) [reply RecoVertex/BeamSpotProducer/plugins/BeamSpotAnalyzer.cc:12]: for the record, this is fixed at https://github.com/cms-sw/cmssw/pull/37085/commits/83ae0ba51edf6cb43c0cb16771bd748c2770d86f.

**Pull request merged**